### PR TITLE
Recommend careful behavior around push streams

### LINKED ISSUE #4930: Handling of a lost push signal
Current design of server push relies on the client being capable of associating two streams: a stream that triggers push (i.e. the stream on which PUSH_PROMISE was sent) and a stream that conveys the pushed response (i.e. the push stream carrying Push Stream Header). Push ID is used for associating the two streams.

However, there is no mechanism that guarantees the delivery of Push ID for neither of the two streams. Both PUSH_PROMISE and Push Stream Header may get lost when the server resets the stream or when the client requests the server to stop sending.

If either signal is lost, the client fails to associate the information that was received on the other side that was not lost, and ends up retaining the other side forever.

Originally reported by @tatsuhiro-t on the [mailing list thread](https://mailarchive.ietf.org/arch/msg/quic/xL43yic8sSnjbofCpEcVYn0RNlw/). Examples that lead up to this problem can be found in [this email](https://mailarchive.ietf.org/arch/msg/quic/xL43yic8sSnjbofCpEcVYn0RNlw/) and [this](https://mailarchive.ietf.org/arch/msg/quic/zoDiBQ08Pp5AXODzXOjBrNbxksc/).

### PR DESCRIPTION
**Pull request by MikeBishop**
Note that all of these recommendations are SHOULD NOT -- an implementation under memory pressure might choose to discard data aggressively, even at the risk of degrading the connection state.  Closes #4930.

### CODE REVIEW

File: draft-ietf-quic-http.md [hunk 1]
```diff
@@ -1193,6 +1198,10 @@ Push Stream Header {
 ~~~~~~~~~~
 {: #fig-push-stream-header title="Push Stream Header"}
 
+A client SHOULD NOT abort reading on a push stream prior to reading the Push ID,
```
Comment (LPardue) [inline draft-ietf-quic-http.md:1201]:
```suggestion
A client SHOULD NOT abort reading a push stream prior to reading the Push ID,
```

maybe?

**Comment by LPardue** [comment]
Thanks Mike. The change looks fine to me, I'll let others debate the finer details.

Assuming this is the preferred direction, I'd ask that we hold the PR and not merge it until all the process stuff is in hand.

**Review by kazuho — APPROVED** [review]

**Review by martinthomson — COMMENTED** [review]
Thanks.  I think that the first change is the key change here; the latter changes are probably not a good idea.  Instead, I would note that if the client does abort reading (for either request streams or push streams) then this risks loss of any push ID that might have been sent, which might lead to timeouts.

**Comment by MikeBishop** [comment]
@martinthomson, can you expand on why you think the latter changes aren't good?  Noting that subpar things might happen if you abort reading seems like the RFC2119 definition of SHOULD NOT abort reading.  Certainly we can explain the SHOULD NOT more than we currently do.

**Comment by martinthomson** [comment]
Fact is: clients will want to discard responses when the thing that asked for that response goes away.  You are saying that they SHOULD NOT do that.  But they are going to anyway.  I would rather we just say that if you have push enabled and you do that, you might be relying more heavily on your push ID timeouts to ensure that you properly handle missing push information.

**Comment by MikeBishop** [comment]
I'm not saying anything about request streams; I agree that clients will want to discard requests they don't need any more and will potentially be discarding PUSH_PROMISE frames along with them.  The timeout in that case is fine.

What I'm specifically recommending in the text is that endpoints read far enough into a unidirectional stream (1-9 bytes) to discover the push ID and mark it consumed.  If they don't, the server might eventually become unable to push; no timeout will recover from that.

**Review by tatsuhiro-t — APPROVED** [review]

File: draft-ietf-quic-http.md [hunk 2]
```diff
@@ -1132,6 +1133,10 @@ an error code of H3_STREAM_CREATION_ERROR or a reserved error code
 ({{http-error-codes}}), but MUST NOT consider such streams to be a connection
 error of any kind.
 
+As certain stream types can affect connection state, a recipient SHOULD NOT
+discard data from incoming unidirectional streams prior to reading the type
+byte.
```
Comment (afrind) [inline draft-ietf-quic-http.md:1138]: Type can be longer than 1 byte (it's a varint)

File: draft-ietf-quic-http.md [hunk 3]
```diff
@@ -884,7 +884,8 @@ as-yet-unknown Push ID, both the associated client request and the pushed
 request header fields are unknown.  The client can buffer the stream data in
 expectation of the matching PUSH_PROMISE. The client can use stream flow control
 (see {{Section 4.1 of QUIC-TRANSPORT}}) to limit the amount of data a server may
-commit to the pushed stream.
+commit to the pushed stream.  Clients SHOULD discard data from push streams if
```
Comment (afrind) [inline draft-ietf-quic-http.md:887]: Does the client also abort reading the push stream if it is incomplete?

File: draft-ietf-quic-http.md (hunk 2 above)
Comment (afrind) [reply draft-ietf-quic-http.md:1138]: Same question about 'discard data' - is the client expected to read this stream until FIN/RST or can it abort?

**Review by martinthomson — APPROVED** [review]
OK, with the explanation, this works for me.

File: draft-ietf-quic-http.md (hunk 2 above)
Comment (MikeBishop) [reply draft-ietf-quic-http.md:1138]: The client is expected to read until it knows the type of the stream.  At that point, it's free to abort or continue reading based on that information.

File: draft-ietf-quic-http.md (hunk 3 above)
Comment (MikeBishop) [reply draft-ietf-quic-http.md:887]: If what is incomplete?  The Push ID?  The response?

Comment (afrind) [reply draft-ietf-quic-http.md:887]: I think the PR is clear that a client needs to process any push stream at least as far as the push ID.  

The new text says the client should 'discard data' if the matching PUSH_PROMISE doesn't arrive in a reasonable time.  Should the client also send a STOP_SENDING in this case?  By incomplete I meant 'had received the Push ID but not the FIN'

Comment (MikeBishop) [reply draft-ietf-quic-http.md]: Ah, I see.  Thanks for clarifying!  I think 0ce850eb addresses this better.

File: draft-ietf-quic-http.md (hunk 1 above)
Comment (LPardue) [reply draft-ietf-quic-http.md:1202]: Other parts of the doc use the style "reading on", so I'll remove this suggestion.

**Review by LPardue — APPROVED** [review]

**Pull request merged**